### PR TITLE
feat(pubsub) Extend PubSub API with on-demand pubslish functionality

### PR DESCRIPTION
**Pull request by andreasebner**
The existing open62541 library only permits data to be published at regular recuring intervals. Use cases exist that need to publish messages on demand. For instance, it is critical to be able to publish in response to a change in data value of a Variable Node.

This API extends the current PubSub implementation with the on-demand publish feature and the capability to get the last publish-timestamp.

### CODE REVIEW

File: src/pubsub/ua_pubsub_writergroup.c [hunk 1]
```diff
@@ -595,6 +595,28 @@ UA_Server_WriterGroup_getState(UA_Server *server, UA_NodeId writerGroupIdentifie
...
+    UA_WriterGroup_publishCallback(server, writerGroup);
+}
+
+UA_DateTime UA_EXPORT
+UA_WriterGroup_lastPublishTimestamp(UA_Server *server, const UA_NodeId writerGroupId){
+    //search WriterGroup ToDo create lookup table for more efficiency
+    UA_WriterGroup *writerGroup;
+    writerGroup = UA_WriterGroup_findWGbyId(server, writerGroupId);
+    if(writerGroup == NULL){
+        return UA_STATUSCODE_BADNOTFOUND;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_writergroup.c:615]: Hmm.
Users of the method will not expect that DateTime can be a StatusCode.

**Comment by applea9** [comment]
I think this PR is very useful. 

From my understanding this public api may be used in the multithreading scenario. Should we protect the low layer operation with lock when regular recuring intervals task has conficts with on-demand pubslish task?

**Comment by jpfr** [comment]
Yes, public API methods should protect against race conditions with `UA_LOCK(&server->serviceMutex)`.
This is already done for the existing methods in that file.

File: src/pubsub/ua_pubsub_writergroup.c [hunk 2]
```diff
@@ -594,6 +594,35 @@ UA_Server_WriterGroup_getState(UA_Server *server, UA_NodeId writerGroupIdentifie
...
 
+UA_StatusCode
+UA_Server_WriterGroup_publish(UA_Server *server, const UA_NodeId writerGroupIdentifier){
+    UA_LOCK(&server->serviceMutex);
+
+    //search WriterGroup ToDo create lookup table for more efficiency
+    UA_WriterGroup *writerGroup;
+    writerGroup = UA_WriterGroup_findWGbyId(server, writerGroupIdentifier);
+    if(writerGroup == NULL){
+        return UA_STATUSCODE_BADNOTFOUND;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_writergroup.c:605]: Unlock missing in Return.

File: src/pubsub/ua_pubsub_writergroup.c [hunk 3]
```diff
@@ -594,6 +594,35 @@ UA_Server_WriterGroup_getState(UA_Server *server, UA_NodeId writerGroupIdentifie
...
+UA_WriterGroup_lastPublishTimestamp(UA_Server *server, const UA_NodeId writerGroupId, UA_DateTime *timestamp){
+    UA_LOCK(&server->serviceMutex);
+    //search WriterGroup ToDo create lookup table for more efficiency
+    UA_WriterGroup *writerGroup;
+    writerGroup = UA_WriterGroup_findWGbyId(server, writerGroupId);
+    if(writerGroup == NULL){
+        return UA_STATUSCODE_BADNOTFOUND;
+    }
+    *timestamp = writerGroup->lastPublishTimeStamp;
+    return UA_STATUSCODE_BADNOTFOUND;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_writergroup.c:622]: You canot unlock After the Return.

File: src/pubsub/ua_pubsub_writergroup.c [hunk 4]
```diff
@@ -594,6 +594,35 @@ UA_Server_WriterGroup_getState(UA_Server *server, UA_NodeId writerGroupIdentifie
...
+}
+
+UA_StatusCode
+UA_WriterGroup_lastPublishTimestamp(UA_Server *server, const UA_NodeId writerGroupId, UA_DateTime *timestamp){
+    UA_LOCK(&server->serviceMutex);
+    //search WriterGroup ToDo create lookup table for more efficiency
+    UA_WriterGroup *writerGroup;
+    writerGroup = UA_WriterGroup_findWGbyId(server, writerGroupId);
+    if(writerGroup == NULL){
+        return UA_STATUSCODE_BADNOTFOUND;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_writergroup.c:619]: Unlock in return.

File: src/pubsub/ua_pubsub_manager.h [hunk 5]
```diff
@@ -36,6 +36,13 @@ typedef struct UA_ReserveId {
     LIST_ENTRY(UA_ReserveId) listEntry;
 } UA_ReserveId;
 
+/*
+ * typedef struct UA_WriterGroup_Lookup {
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_manager.h:40]: Hum?

File: include/open62541/server_pubsub.h [hunk 6]
```diff
@@ -549,6 +549,13 @@ UA_StatusCode UA_EXPORT
 UA_Server_WriterGroup_getState(UA_Server *server, UA_NodeId writerGroupIdentifier,
                                UA_PubSubState *state);
 
+UA_StatusCode UA_EXPORT
```
Comment (jpfr) [inline include/open62541/server_pubsub.h:552]: Add threadsafe.

**Review by jpfr — COMMENTED** [review]
Nice and tight.
Nearly ready to merge.

**Pull request merged**